### ISSUE TITLE
Support AOT/Trimming/Blazor WebAssembly using source generation

### ISSUE THREAD

**Issue by cyrmax** [TelegramBots/Telegram.Bot#1322]
## I would like to request/propose a feature

I suggest replacing NewtonSoft.Json with System.Text.Json with proper source generation used.
The reason is that with newtonsoft we cannot use Aot and/or trimming which then breaks the entire Json package.
With System.Text.Json we can see the similar issue with trimming and Aot. But there we can use source generation aproach and avoid any errors when trimming the executable or building native code.
Here is the documentation for this source generation feature in detail and System.Text.Json docs can be googled easily.
https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/source-generation?pivots=dotnet-8-0

If anybody has any arguments against my feature request I would be glad to discuss them.
Also I will try to implement the thing I want by myself and if I achieve what I want without any errors I of course will make a pull request to address this issue.

Thanks in advance.

**Comment by karb0f0s** [comment]
@cyrmax wow, what a great idea! How would you approach it with current codebase?

**Comment by tuscen** [comment]
Will this work for `netstandard2.0` targets?

**Comment by cyrmax** [comment]
@tuscen 
I found the following information regarding your question.
TLDR: yes.
But here is a quote and a link.

> The System.Text.Json library is included in the runtime for .NET Core 3.1 and later versions. For other target frameworks, install > the System.Text.Json NuGet package. The package supports:
> • .NET Standard 2.0 and later versions
> • .NET Framework 4.7.2 and later versions
> • .NET Core 2.0, 2.1, and 2.2

And here is the link:
https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/migrate-from-newtonsoft?pivots=dotnet-8-0

**Comment by tuscen** [comment]
@cyrmax Thanks, we'll probably try to migrate to source generated converters then

**Comment by cyrmax** [comment]
@karb0f0s 
If you mean that it is impossible or too difficult then please explain why.
I didn't looked at the code yet but I think at least it is possible and reasonable.

**Comment by tuscen** [comment]
@cyrmax By the way, you really don't want this library to be trimmed since it's an API client. What will it do in case your bot receives a type that was trimmed during compilation?

**Comment by tuscen** [comment]
@cyrmax Do you have experience with STJ source generators? I’ve tried to use it in this library, but for some reason it doesn’t generate converters. But in a separate test project I don’t have any problems.

**Comment by cyrmax** [comment]
@tuscen I will try to investigate this on weekend. In my projects this feature works just fine too. I will see.

**Comment by tarasverq** [comment]
> I've managed to generate json converters, but I struggle to understand how to make it work in .NET 6 in such a way that one could augment existing `JsonSerializerOptions` instead of completely rewriting existing ones. This is a deal breaker since we can't impose our own instance of `JsonSerializerOptions` on clients using webhooks since they might have their own conventions in their ASP.NET Core apps. It seems .NET 8 have a solution for this using `JsonTypeInfoResolver.Combine` which .NET 6 lacks. If you have ideas on how to workaround please tell me. Otherwise I don't think we will be able to use STJ until .NET 6 is EOL so we could target the library to .NET 8 instead of .NET 6.

https://www.planetgeek.ch/2022/10/15/using-system-text-json-alongside-newtonsoft-json-net/
In this article author shows how to use System.Text.Json alongside Newtonsoft.
And also he shows how it's possible to change serializer options for endpoints marked with special attribute (file `SystemTextJsonBodyModelBinder.cs` lines 10-11)

Maybe it would be useful for further research.

**Comment by wiz0u** [comment]
We are now using System.Text.Json in recent versions of the library.
I keep this issue open to follow up on AOT/Trimming/Source generation compatiblity

**Comment by wiz0u** [comment]
Should be supported in 22.2.0

**Comment by dvpot** [comment]
Hi
I have the same problem as [#1417](https://github.com/TelegramBots/Telegram.Bot/issues/1417)

Errors appear after installing the Telegram.Bot package.

![Image](https://github.com/user-attachments/assets/30bbab0e-90f9-463a-af45-9270985b7679)

**Comment by nguyenngocanh94** [comment]
any suggestion for me, I already use the version 22.5.0-dev.12 but still can not use  it for native AOT asp.net
throw error when build.

obj/Debug/net9.0/System.Text.Json.SourceGeneration/System.Text.Json.SourceGeneration.JsonSourceGenerator/TelegramJsonSerializerContext.PaidMediaArray.g.cs(57,85): error CS0103: The name 'PaidMedia' does not exist in the current context

already declared the json:

[JsonSerializable(typeof(PaidMedia[]))]
[JsonSerializable(typeof(Message))]
[JsonSerializable(typeof(User))]
[JsonSerializable(typeof(Chat))]
internal partial class TelegramJsonSerializerContext : JsonSerializerContext
{
    
}

**Comment by wiz0u** [comment]
Suggestion: use the nuget, not the source. And call Services.ConfigureTelegramBot so you don't need your own context
